src/trie.rs

```diff
@@ -60,8 +60,8 @@ where
             self.remove(key)?;
             return Ok(());
         }
-        let root = self.insert_at(self.root.clone(), &Nibbles::from_raw(key, true), value)?;
-        self.root = root;
+        let root = self.root.clone();
```

Comment:
this is the only one `clone`. note we also can avoid this, if `insert_at` return the param node when error encountered like:
```rust
type TrieInsertResult<T, C, D> = Result<T, (Node, TrieError<C, D>)>;
fn insert_at(&mut self, n: Node, partial: &Nibbles, value: &[u8]) -> TrieInsertResult<Node, C, D> ;
```
```rust
    let root = self.root.take();
    self.root = self.insert_at(root, &Nibbles::from_raw(key, true), value).map_err(|(root, e)| {
        self.root = root;
        e
    })?;
```

Reply:
Is there any problem with the code below?
```rust
  let root = self.root.take();
  self.root = self.insert_at(root, &Nibbles::from_raw(key, true), value)?;
```

Reply:
if error occurs, we should guarantee `self.root` is logically unchanged, instead making it an empty trie.

Reply:
i see, thanks